src/peergos/shared/user/fs/TagIndexingService.java

```diff
@@ -0,0 +1,95 @@
+    };
+    public static final String TAGS_INDEX_FILENAME = ".tagsindex";
+
+    private final FileWrapperUpdater tagsIndexDirUpdater;
+    private final FileWrapper tagsIndexHomeDir;
+    private final TransactionService transactions;
+    private final NetworkAccess networkAccess;
+    private final Crypto crypto;
+
+    public TagIndexingService(NetworkAccess networkAccess,
+                              TransactionService transactions,
+                              Crypto crypto,
+                              FileWrapper tagsIndexDir) {
+        this.tagsIndexDirUpdater = v -> tagsIndexDir.getUpdated(v, networkAccess);
+        this.tagsIndexHomeDir = tagsIndexDir;
+        this.transactions = transactions;
+        this.networkAccess = networkAccess;
+        this.crypto = crypto;
+    }
+
+    public CompletableFuture<TagIndex> addTagIndexEntries(TagIndex tagIndex, Map<String, Set<String>> entries) {
+        return CompletableFuture.completedFuture(tagIndex.addEntries(entries));
+    }
+
+    public CompletableFuture<TagIndex> removeTagIndexEntries(TagIndex tagIndex,  Map<String, Set<String>> entries) {
+        return CompletableFuture.completedFuture(tagIndex.removeEntries(entries));
+    }
+
+    public CompletableFuture<Snapshot> modifyTagIndexEntries(Snapshot version, Committer committer,
+                                                             Map<String, Set<String>> entriesToAdd,
+                                                             Map<String, Set<String>> entriesToRemove) {
+        return getTagIndexEntries(version).thenCompose(originalTagIndex -> removeTagIndexEntries(originalTagIndex, entriesToRemove))
+                .thenCompose(tagIndexAfterRemoval -> addTagIndexEntries(tagIndexAfterRemoval, entriesToAdd))
+                .thenCompose(tagIndexAfterAdding -> modifyTagIndexEntries(version, committer, tagIndexAfterAdding));
+    }
+
+    private CompletableFuture<Snapshot> modifyTagIndexEntries(Snapshot s, Committer committer, TagIndex newIndex) {
+        byte[] data = newIndex.serialize();
+        AsyncReader asyncReader = AsyncReader.build(data);
+
+        return tagsIndexHomeDir.getPath(networkAccess).thenCompose(path ->
+                Transaction.buildFileUploadTransaction(Paths.get(path).resolve(TAGS_INDEX_FILENAME).toString(), data.length,
+                        asyncReader, tagsIndexHomeDir.signingPair(),
+                        tagsIndexHomeDir.generateChildLocationsFromSize(data.length, crypto.random)))
+                .thenCompose(txn -> transactions.open(s, committer, txn).thenCompose(v -> asyncReader.reset()
+                                .thenCompose(reset -> tagsIndexHomeDir.uploadFileSection(v, committer,
+                                        TAGS_INDEX_FILENAME, null, reset, false, 0,
+                                        data.length, Optional.empty(), true, true, networkAccess, crypto,
+                                        x -> {}, txn.getLocations().get(0).getMapKey(),
```

Comment:
🟢 
Small thing, but couldn't we use `VOID_PROGRESS` instead of `x -> {}`?

Reply:
We definitely could use VOID_PROGRESS - done :)